src/ring_theory/ore_localization/basic.lean

```diff
@@ -0,0 +1,785 @@
+
+@[simp]
+protected lemma div_eq_one {s : S} : (s : R) /ₒ s = 1 :=
+by { cases s; apply ore_localization.div_eq_one' }
+
+protected lemma one_mul (x : R[S ⁻¹]) : 1 * x = x :=
+begin
+  induction x using ore_localization.ind with r s,
+  simp [ore_localization.one_def, ore_div_mul_char (1 : R) r (1 : S) s r 1 (by simp)]
+end
+
+protected lemma mul_one (x : R[S ⁻¹]) : x * 1 = x :=
+begin
+  induction x using ore_localization.ind with r s,
+  simp [ore_localization.one_def, ore_div_mul_char r 1 s 1 1 s (by simp)]
+end
+
+protected lemma mul_assoc (x y z : R[S ⁻¹]) : x * y * z = x * (y * z) :=
+begin
+  induction x using ore_localization.ind with r₁ s₁,
+  induction y using ore_localization.ind with r₂ s₂,
+  induction z using ore_localization.ind with r₃ s₃,
+  rcases ore_div_mul_char' r₁ r₂ s₁ s₂ with ⟨ra, sa, ha, ha'⟩, rw ha', clear ha',
+  rcases ore_div_mul_char' r₂ r₃ s₂ s₃ with ⟨rb, sb, hb, hb'⟩, rw hb', clear hb',
+  rcases ore_condition rb sa with ⟨rc, sc, hc⟩,
+  rw ore_div_mul_char (r₁ * ra) r₃ (s₂ * sa) s₃ rc (sb * sc)
+    (by { simp only [submonoid.coe_mul], assoc_rw [hb, hc] }),
+  rw [mul_assoc, ←mul_assoc s₃],
+  symmetry, apply ore_div_mul_char,
+  assoc_rw [hc, ←ha], apply mul_assoc
+end
+
+instance : monoid R[S⁻¹] :=
+{ one_mul := ore_localization.one_mul,
+  mul_one := ore_localization.mul_one,
+  mul_assoc := ore_localization.mul_assoc,
+  .. ore_localization.has_mul,
+  .. ore_localization.has_one }
+
+protected lemma mul_inv (s s' : S) : ((s : R) /ₒ s') * (s' /ₒ s) = 1 :=
+by simp [ore_div_mul_char (s :R) s' s' s 1 1 (by simp)]
+
+@[simp]
+protected lemma mul_one_div {r : R} {s t : S} : (r /ₒ s) * (1 /ₒ t) = r /ₒ (t * s) :=
+by simp [ore_div_mul_char r 1 s t 1 s (by simp)]
+
+@[simp]
+protected lemma mul_cancel {r : R} {s t : S} : (r /ₒ s) * (s /ₒ t) = r /ₒ t :=
+by simp [ore_div_mul_char r s s t 1 1 (by simp)]
+
+@[simp]
+protected lemma mul_cancel' {r₁ r₂ : R} {s t : S} : (r₁ /ₒ s) * ((s * r₂) /ₒ t) = (r₁ * r₂) /ₒ t :=
+by simp [ore_div_mul_char r₁ (s * r₂) s t r₂ 1 (by simp)]
+
+@[simp]
+lemma div_one_mul {p r : R} {s : S} :
+  (r /ₒ 1) * (p /ₒ s) = (r * p) /ₒ s := --TODO use coercion r ↦ r /ₒ 1
+by simp [ore_div_mul_char r p 1 s p 1 (by simp)]
+
+/-- The unit in `R[S⁻¹]` consisting of the fraction `s /ₒ 1` for `s : S`. -/
+def numerator_unit (s : S) : units R[S⁻¹] :=
+{ val := (s : R) /ₒ 1,
+  inv := (1 : R) /ₒ s,
+  val_inv := ore_localization.mul_inv s 1,
+  inv_val := ore_localization.mul_inv 1 s }
+
+/-- The multiplicative homomorphism from `R` to `R[S⁻¹]`, mapping `r : R` to the
+fraction `r /ₒ 1`. -/
+def numerator_hom : R →* R[S⁻¹] :=
+{ to_fun := λ r, r /ₒ 1,
+  map_one' := rfl,
+  map_mul' := λ r₁ r₂, div_one_mul.symm }
+
+lemma numerator_hom_apply {r : R} : numerator_hom r = r /ₒ (1 : S) := rfl
+
+lemma numerator_is_unit (s : S) : is_unit ((numerator_hom (s : R)) : R[S ⁻¹]) :=
+⟨numerator_unit s, rfl⟩
+
+section UMP
+
+variables {T : Type*} [monoid T]
+variables (f : R →* T) (fS : S →* units T)
```

Comment:
(and, it's the only case where I'd have to slap a `noncomputable` badge on a definition)

Reply:
I think `(hf s).unit⁻¹` is good enough a notation. It might be a nice experiment to do things differently (computably) in the noncommutative case, and I'll leave that to maintainers to decide.

Reply:
If you keep the current approach, I think it suffices for `fS` to be just a function `(fS : S → units T)` instead of a monoid_hom.